File: src/adapters/openai.ts
/**
 * OpenAI embedder adapter.
 * Requires: npm install openai
 *
 * Usage:
 *   import { AgentMemory } from 'agentic-memory';
 *   import { OpenAIEmbedder } from 'agentic-memory/adapters/openai';
 *
 *   const memory = new AgentMemory({
 *     embedder: new OpenAIEmbedder({ apiKey: process.env.OPENAI_API_KEY }),
 *   });
 */

import type { Embedder } from '../core/types.js';

export interface OpenAIEmbedderConfig {
  apiKey: string;
  model?: string;
  dimensions?: number;
  baseURL?: string;
}

export class OpenAIEmbedder implements Embedder {
  private apiKey: string;
  private model: string;
  private dim: number;
  private baseURL: string;

  constructor(config: OpenAIEmbedderConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model ?? 'text-embedding-3-small';
    this.dim = config.dimensions ?? 1536;
    this.baseURL = config.baseURL ?? 'https://api.openai.com/v1';
  }

  dimensions(): number {
    return this.dim;
  }

  async embed(text: string): Promise<number[]> {
    const res = await fetch(`${this.baseURL}/embeddings`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: text,
        dimensions: this.dim,
      }),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`OpenAI embedding failed (${res.status}): ${err}`);
    }

    const data = await res.json() as { data: { embedding: number[] }[] };
    return data.data[0].embedding;
  }

  async embedBatch(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) return [];

    const res = await fetch(`${this.baseURL}/embeddings`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: texts,
        dimensions: this.dim,
      }),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`OpenAI embedding failed (${res.status}): ${err}`);
    }

    const data = await res.json() as { data: { embedding: number[]; index: number }[] };
    // OpenAI may return out of order, sort by index
    return data.data
      .sort((a, b) => a.index - b.index)
      .map(d => d.embedding);
  }
}

File: python/agentic_memory/adapters/openai.py
"""
OpenAI embedder adapter.
Requires: pip install httpx (or aiohttp)

Usage:
    from agentic_memory import AgentMemory
    from agentic_memory.adapters import OpenAIEmbedder

    memory = AgentMemory(
        embedder=OpenAIEmbedder(api_key="sk-...")
    )
"""

from __future__ import annotations

import json
from urllib.request import Request, urlopen
from typing import Optional


class OpenAIEmbedder:
    def __init__(
        self,
        api_key: str,
        model: str = "text-embedding-3-small",
        dim: int = 1536,
        base_url: str = "https://api.openai.com/v1",
    ) -> None:
        self._api_key = api_key
        self._model = model
        self._dim = dim
        self._base_url = base_url

    def dimensions(self) -> int:
        return self._dim

    async def embed(self, text: str) -> list[float]:
        result = self._call([text])
        return result[0]

    async def embed_batch(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []
        return self._call(texts)

    def _call(self, input_texts: list[str]) -> list[list[float]]:
        payload = json.dumps({
            "model": self._model,
            "input": input_texts,
            "dimensions": self._dim,
        }).encode("utf-8")

        req = Request(
            f"{self._base_url}/embeddings",
            data=payload,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._api_key}",
            },
            method="POST",
        )

        with urlopen(req) as resp:
            if resp.status != 200:
                raise RuntimeError(f"OpenAI embedding failed ({resp.status})")
            data = json.loads(resp.read())

        # Sort by index (OpenAI may return out of order)
        items = sorted(data["data"], key=lambda d: d["index"])
        return [d["embedding"] for d in items]
